File: Examenes-EDD-5/2019-Febrero/Haskell/Kruskal.hs
----------------------------------------------
-- Estructuras de Datos.  2018/19
-- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
-- Escuela Técnica Superior de Ingeniería en Informática. UMA
--
-- Examen 4 de febrero de 2019
--
-- ALUMNO/NAME: Paco Molina Cuenca
-- GRADO/STUDIES: Doble Grado Matemáticas e Ingeniería Infomática
-- NÚM. MÁQUINA/MACHINE NUMBER:
--
----------------------------------------------

module Kruskal(kruskal, kruskals) where

import qualified DataStructures.Dictionary.AVLDictionary as D
import qualified DataStructures.PriorityQueue.LinearPriorityQueue as Q
import DataStructures.Graph.DictionaryWeightedGraph

kruskal :: (Ord a, Ord w) => WeightedGraph a w -> [WeightedEdge a w]
kruskal g = kruskalRec (foldr (\v -> D.insert v v) D.empty (vertices g)) (foldr Q.enqueue Q.empty (edges g)) []
    where 
        kruskalRec dict pq t 
            | Q.isEmpty pq = t 
            | otherwise    = if repOrg /= repDst then kruskalRec (D.updateOrInsert repDst (const org) org dict) (Q.dequeue pq) (edge:t) else kruskalRec dict (Q.dequeue pq) t
                where
                    edge@(WE org w dst) = Q.first pq  
                    repOrg = representante org 
                    repDst = representante dst

                    representante x =
                        case D.valueOf x dict of 
                            Nothing -> error ""
                            Just y  -> if x == y then x else representante y

-- Solo para evaluación continua / only for part time students
kruskals :: (Ord a, Ord w) => WeightedGraph a w -> [[WeightedEdge a w]]
kruskals = undefined

File: Examenes-EDD-6/2019-Febrero/Java/Kruskal.java
/**----------------------------------------------
 * -- Estructuras de Datos.  2018/19
 * -- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
 * -- Escuela Técnica Superior de Ingeniería en Informática. UMA
 * --
 * -- Examen 4 de febrero de 2019
 * --
 * -- ALUMNO/NAME: Paco Molina Cuenca
 * -- GRADO/STUDIES: Doble Grado Matemáticas e Ingeniería Informática
 * -- NÚM. MÁQUINA/MACHINE NUMBER:
 * --
 * ----------------------------------------------
 */

import dataStructures.dictionary.Dictionary;
import dataStructures.dictionary.HashDictionary;
import dataStructures.graph.WeightedGraph;
import dataStructures.graph.WeightedGraph.WeightedEdge;
import dataStructures.priorityQueue.LinkedPriorityQueue;
import dataStructures.priorityQueue.PriorityQueue;
import dataStructures.set.HashSet;
import dataStructures.set.Set;

public class Kruskal {
	public static <V,W> Set<WeightedEdge<V,W>> kruskal(WeightedGraph<V,W> g){
		Dictionary<V,V> dict = new HashDictionary<>();
		
		for(V v : g.vertices())
			dict.insert(v, v);

		PriorityQueue<WeightedEdge<V,W>> pq = new LinkedPriorityQueue<>();

		for(WeightedEdge<V,W> edge : g.edges())
			pq.enqueue(edge);

		Set<WeightedEdge<V,W>> t = new HashSet<>();

		WeightedEdge<V,W> edge;

		V repOrg, repDst;

		while(!pq.isEmpty()){
			edge = pq.first();
			pq.dequeue();

			repOrg = representante(edge.source(), dict);
			repDst = representante(edge.destination(), dict);

			if(!repOrg.equals(repDst)){
				dict.insert(repDst, edge.source());
				t.insert(edge);
			}
		}

		return t;
	}

	private static <V> V representante(V v, Dictionary<V,V> dict){
		return dict.valueOf(v).equals(v) ? v : representante(dict.valueOf(v), dict);
	}

	// Sólo para evaluación continua / only for part time students
	public static <V,W> Set<Set<WeightedEdge<V,W>>> kruskals(WeightedGraph<V,W> g) {

		// COMPLETAR
		
		return null;
	}
}
